server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java

```diff
@@ -221,12 +225,13 @@ public float score() throws IOException {
             int docId = docID();
             scoreScript.setDocument(docId);
             float score = (float) scoreScript.execute(explanation);
-            if (score == Float.NEGATIVE_INFINITY || Float.isNaN(score)) {
-                throw new ElasticsearchException(
-                    "script_score query returned an invalid score [" + score + "] for doc [" + docId + "].");
+            if (score < 0f || Float.isNaN(score)) {
```

Comment:
It looks like this PR also fixed a bug in `script_score` queries where we allowed negative scores. I think we should add a note to the breaking changes docs and also update the PR description to make it clear we included this change.

Reply:
@jtibshirani Thanks.  Even before without this change a user would get an error if their `script_score` query produced a negative score. They would just get it from a different place, I think from the Lucene [here](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79)

So the only thing changed from a user perspective is an error message and error status code (before was 500, not 400x).  Do you think it warrants a breaking change notice?
+1 for include this in the PR description

Reply:
I tried this out using an Elasticsearch 7.6 build, and didn't receive an error:

```
PUT my_index/_doc/1?refresh
{
  "field": "value"
}

GET my_index/_search
{
  "query": {
    "script_score": {
      "query": {
        "match_all": {}
      },
      "script": {
        "source": "-1000"
      }
    }
  }
}
```

The line you linked to is an assert, so perhaps these Lucene checks didn't always catch the issue in non-test environments.

Reply:
@jtibshirani Thanks for uncovering this.  I understood what happened:
- Before 7.5,  `script_score` query was using [ScriptScoreFunction](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreFunction.java#L83) that was returning 400 error with a negative score.  
- From 7.5, we have changed it to not use `ScriptScoreFunction` but forgot to add a condition for a negative score.  But  [TopScoreDocCollector assertion](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java#L79) is tripped, causing fatal error in the dev mode. But I guess we silence these assertions in a production mode  as we don't see any visible errors or error log messages.

So, negative scores were wrongly allowed only in 7.5-7.6 versions, so to me it doesn't look like  a breaking change. WDYT?

Reply:
Adding an explanation to the release notes makes sense to me. I agree it shouldn't be presented as a typical 'breaking change', it is more like a regression that we fixed. Perhaps we could add a unit test along with the release notes update, to prevent a future regression?